### ISSUE TITLE
Default link name of items in extern blocks is not documented

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang-nursery/reference#658]
The extern block section does not document the "default" `link_name` used for items (functions and statics) in extern blocks. @RalfJung shows here (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f976fe163c70645529ae6e1b15fd82a) that independently of the ABI used the unmangled name is chosen, which makes sense as the intended behavior (what else could we do?). cc @eddyb

**Comment by eddyb** [comment]
Yeah, `#[link_name]` overrides the name of the import itself.

**Comment by gnzlbg** [comment]
It's kind of weird that, e.g. `extern "C" { fn foo(); }` gives `foo` the unmangled link name `foo`, but that `extern "C" fn foo() { ... }` requires `#[no_mangle]`.

**Comment by eddyb** [comment]
"Maybe we shouldn't have used the keyword `extern` for modifying the ABI of a function" strikes again.

**Comment by gnzlbg** [comment]
So why do extern blocks change mangling ? Is this always to the unmangled name ? E.g. for `extern "Rust" { fn foo(); }` it might have made sense to use the Rust mangling scheme for `foo` - i'm not sure, is probably a backward incompatible change at this point.

**Comment by eddyb** [comment]
> Is this always to the unmangled name?

Yeah, Rust-specific mangling is only used for non-FFI definitions, and the compiler has knowledge of the definition, that's the only reason the mangling *makes sense*.

`extern "Rust" { fn foo(); }` couldn't use any mangling *even if we wanted to*. There's not enough information there (i.e. crate and crate disambiguator hash) to generate a mangled name

**Comment by gnzlbg** [comment]
Yeah, there is only one obvious way to do this. We should just write this down somewhere.
